### PR TITLE
ICU-22365 C API for ULocaleBuilder

### PR DESCRIPTION
**Pull request by FrankYFTang**
Still need to add tests

##### Checklist

- [x] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-22365
- [X] Required: The PR title must be prefixed with a JIRA Issue number. 
- [X] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [X] Required: Each commit message must be prefixed with a JIRA Issue number. 
- [X] Issue accepted (done by Technical Committee after discussion)
- [ ] Tests included, if applicable
- [X] API docs and/or User Guide docs changed or added, if applicable

### CODE REVIEW

File: icu4c/source/common/ulocbuilder.cpp
Comment (richgillam) [inline icu4c/source/common/ulocbuilder.cpp:61]: Didn't @markusicu say something this morning about StringPiece having logic on it that'd handle checking for the negative length for you?  Or am I wrong about this?

File: icu4c/source/common/ulocbuilder.cpp
Comment (richgillam) [inline icu4c/source/common/ulocbuilder.cpp:141]: I actually like this, but didn't we have a fairly big argument in ICU TC not that long about about my proposal to make the `uloc_getXXX()` methods always null-terminate their result buffers?  I guess there's a backward-compatibility issue with doing it to those methods (although it doesn't actually seem that bad to me), but the big issue seemed to be that we've always defined all these methods that fill in buffers as _not_ null-terminating the buffer if the result's length equals or exceeds the buffer length.  Is it okay for this method to behave differently?  Do we need to re-litigate that discussion?  @markusicu

File: icu4c/source/common/unicode/ulocbuilder.h
Comment (richgillam) [inline icu4c/source/common/unicode/ulocbuilder.h:29]: "string of locale" sounds funny to me.  Maybe "locale ID string"?

File: icu4c/source/common/unicode/ulocbuilder.h
Comment (richgillam) [inline icu4c/source/common/unicode/ulocbuilder.h:70]: s/destoried/destroyed

File: icu4c/source/common/unicode/ulocbuilder.h
Comment (richgillam) [inline icu4c/source/common/unicode/ulocbuilder.h:132]: Syntactically well-formed, but not necessarily valid, right?

Does syntactic well-formedness extend to case, or is the logic case-insensitive?

File: icu4c/source/common/unicode/ulocbuilder.h
Comment (richgillam) [inline icu4c/source/common/unicode/ulocbuilder.h:276]: Dumb question: What's a "Unicode locale attribute"?

File: icu4c/source/test/cintltst/ulocbuildertst.c
Comment (richgillam) [inline icu4c/source/test/cintltst/ulocbuildertst.c:344]: Why are all these tests commented out?

File: icu4c/source/test/cintltst/ulocbuildertst.c
Comment (richgillam) [inline icu4c/source/test/cintltst/ulocbuildertst.c:1134]: Why are these tests commented out?

File: icu4c/source/test/cintltst/ulocbuildertst.c
Comment (richgillam) [inline icu4c/source/test/cintltst/ulocbuildertst.c:1203]: Why commented out?

File: icu4c/source/test/cintltst/ulocbuildertst.c
Comment (richgillam) [inline icu4c/source/test/cintltst/ulocbuildertst.c:1699]: Why commented out?

File: icu4c/source/test/cintltst/ulocbuildertst.c
Comment (richgillam) [inline icu4c/source/test/cintltst/ulocbuildertst.c:1742]: Why are so many of these tests commented out?

**Review by richgillam — APPROVED** [review]
Looks good!  A few nits and questions, but I'll go ahead and approve.

File: icu4c/source/common/ulocbuilder.cpp
Comment (FrankYFTang) [reply icu4c/source/common/ulocbuilder.cpp]: The API doc said 
https://unicode-org.github.io/icu-docs/apidoc/dev/icu4c/classicu_1_1StringPiece.html#af97f22b270d21d5c97337d5d58966a59



len | the length of the string; must be non-negative
-- | --


my understanding is if you use the single parameter one it will figure out the length for you by assuming null termination. If you use the one with length  then the length will be used and no need for null termination. I cannot find one as what you remember.

File: icu4c/source/common/ulocbuilder.cpp
Comment (FrankYFTang) [reply icu4c/source/common/ulocbuilder.cpp:103]: sorry, I just have bad code here. This code will cause uprv_memcpy to segfault because bufferCapacity could be much biggger than the memory allocated for l.getName(). The code in this PR is not ready for review yet since I am still waiting for the API to be approved by the ICU TC and I have not yet complete the testing code.

File: icu4c/source/common/unicode/ulocbuilder.h
Comment (FrankYFTang) [reply icu4c/source/common/unicode/ulocbuilder.h:132]: YES / the syntax is case-insenstive per https://unicode.org/reports/tr35/#Unicode_language_identifier

File: icu4c/source/test/cintltst/ulocbuildertst.c
Comment (FrankYFTang) [reply icu4c/source/test/cintltst/ulocbuildertst.c:344]: I am still working on them.

File: icu4c/source/test/cintltst/ulocbuildertst.c
Comment (FrankYFTang) [reply icu4c/source/test/cintltst/ulocbuildertst.c:1203]: I am still working on them since the API is not yet approved.

File: icu4c/source/test/cintltst/ulocbuildertst.c
Comment (FrankYFTang) [reply icu4c/source/test/cintltst/ulocbuildertst.c:1699]: I am still working on them since the API is not yet approved.

File: icu4c/source/test/cintltst/ulocbuildertst.c
Comment (FrankYFTang) [reply icu4c/source/test/cintltst/ulocbuildertst.c:1742]: I am still working on them since the API is not yet approved so this PR (except the header part) is not ready for review yet.

File: icu4c/source/common/ulocbuilder.cpp
Comment (richgillam) [reply icu4c/source/common/ulocbuilder.cpp]: Bleah.  That's unfortunate.  But okay; I just wanted to make sure.

**Review by richgillam — COMMENTED** [review]
Started to review this again, but it looks like some of the stuff you said wasn't done yet is still in there, so I guess it's not actually ready for me to review again?

File: icu4c/source/test/cintltst/ulocbuildertst.c
Comment (FrankYFTang) [reply icu4c/source/test/cintltst/ulocbuildertst.c]: ready to review now.

File: icu4c/source/test/cintltst/ulocbuildertst.c
Comment (FrankYFTang) [reply icu4c/source/test/cintltst/ulocbuildertst.c]: ready to review now.

File: icu4c/source/test/cintltst/ulocbuildertst.c
Comment (FrankYFTang) [reply icu4c/source/test/cintltst/ulocbuildertst.c]: ready to review now.

**Comment by FrankYFTang** [comment]
Sorry, now it is ready to review.

File: icu4c/source/common/unicode/ulocbuilder.h
Comment (FrankYFTang) [reply icu4c/source/common/unicode/ulocbuilder.h:276]: "attribute = alphanum{3,8} ;	" in unicode_locale_extensions
https://www.unicode.org/reports/tr35/tr35.html#unicode_locale_extensions

A 'u' extension may contain multiple attribute s or keyword s as defined in [Unicode locale identifier](https://www.unicode.org/reports/tr35/tr35.html#Unicode_locale_identifier). The canonical syntax is defined as in [Canonical Unicode Locale Identifiers](https://www.unicode.org/reports/tr35/tr35.html#Canonical_Unicode_Locale_Identifiers).

Also see 
https://unicode-org.github.io/icu-docs/apidoc/released/icu4j/com/ibm/icu/util/ULocale.Builder.html#addUnicodeLocaleAttribute-java.lang.String-

https://download.java.net/java/early_access/jdk21/docs//api/java.base/java/util/Locale.Builder.html#addUnicodeLocaleAttribute(java.lang.String)

**Review by richgillam — APPROVED** [review]
Looks great.

**Comment by FrankYFTang** [comment]
I corrected the comment per @garywade comments on the API proposal. The  additional build method will be added in a separate PR, after the landing of this and the C API for ULocale proposal PR.

**Comment by FrankYFTang** [comment]
@richgillam @garywade - need your final review again. Thanks

**Review by richgillam — APPROVED** [review]
Rubber-stamp approval.  I can't see what you changed since the last time I looked at this (maybe wait to squash until everything's approved?), so I'm trusting you.

File: icu4c/source/common/ulocbuilder.cpp
Comment (markusicu) [inline icu4c/source/common/ulocbuilder.cpp:108]: @FrankYFTang @richgillam sorry for not paying attention for way too long...
Any C API that writes a string to a destination buffer should return with a call to [u_terminateChars()](https://github.com/unicode-org/icu/blob/main/icu4c/source/common/ustr_imp.h#L81), u_terminateUChars(), ... as appropriate, and those functions handle the errors and warnings and NUL-termination. See other C API implementations. Please send a follow-up PR.

Comment (FrankYFTang) [reply icu4c/source/common/ulocbuilder.cpp:108]: @markusicu see https://github.com/unicode-org/icu/pull/2565 for the follow-up PR

**Pull request merged**
